cloudwatch_monitoring/sns.py

```diff
@@ -0,0 +1,184 @@
+
+    sns_widgets.append(sns_notifications_widget)
+
+    return sns_widgets
+
+
+def generate_number_of_messages_published_metric(topic_name, tier_agnostic_topic_name, count):
+    try:
+        label = sns_topics[tier_agnostic_topic_name]['title']
+    except KeyError:
+        # No title property or possibly lookup added for this sns topic yet, default to topic name
+        label = topic_name
+
+    if count < 1:
+        # the first metric in the list has some additional stuff
+        metric = ["AWS/SNS", 'NumberOfMessagesPublished', "TopicName", topic_name, {"label": label}]
+    else:
+        metric = ["...", topic_name, {"label": label}]
+
+    return metric
+
+
+class SNSAPICalls:
+    def __init__(self, region, deploy_stage):
+        """
+        Constructor for the SNSAPICalls class.
+
+        :param region: usually 'us-west-2'
+        :param deploy_stage: The deployment tier (DEV, TEST, QA, PROD-EXTERNAL)
+        """
+        self.region = region
+        self.sns_client = boto3.client('sns', region_name=region)
+        self.deploy_stage = deploy_stage
+
+    def get_all_sns_topics(self):
+        """
+        Using the AWS python sdk (boto3), grab all the sns topics for the specified account for a given region.
+
+        :return: response: a page of sns topics in the account.
+        :rtype: dict
+        """
+
+        # TODO maybe get a paginator to work instead of 'manual' iteration
+        # https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sns.html#SNS.Client.list_topics
+        response = {}
+        next_token = None
+        while True:
+            if next_token:
+                response_iterator = self.sns_client.list_topics(
+                    NextToken=next_token)
+                response['Topics'].extend(response_iterator['Topics'])
+            else:
+                response_iterator = self.sns_client.list_topics()
+                response.update(response_iterator)
+            try:
+                next_token = response_iterator['NextToken']
+            except KeyError:
+                # no more pages, move on
+                break
+
+        return response
+
+    def is_iow_topic_filter(self, topic_arn):
+        """
+        Apply filters to determine if the topic is a tagged IOW asset in the correct tier.
+
+        :param topic_arn: A single sns topic arn
+        :return: is_iow_topic: is this an IOW topic or not
+        :rtype: bool
+        """
+        is_iow_topic = False
+
+        # filtering on deploy tier, which we capitalize
+        if self.deploy_stage.upper() in topic_arn:
+            # launch API call to grab the tags for the sns topic
+            # https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sns.html#SNS.Client.list_tags_for_resource
+            topic_tags = self.sns_client.list_tags_for_resource(ResourceArn=topic_arn)
+
+            # we only want topics that are tagged as 'IOW'
+            if 'Tags' in topic_tags:
```

Comment:
looks cyclomatically complex, even if it isn't.  Maybe combine some items on one line?  Just a thought.

Reply:
Yes this is a lot of ifs... similar logic exists in some of the other tag filtering methods for other asset types.  I may create a small refactor task to look at some of these things and make them better.